src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesIterator.java

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * Total number of live doc which will the iterator will iterate upon.
+     * @return long: total number of live docs
+     */
+    long liveDocs();
+
+    /**
+     * A DocIdsIteratorValues provides a common iteration logic for all Values that implements
+     * {@link DocIdSetIterator} interface. Example: {@link BinaryDocValues}, {@link FloatVectorValues} etc.
+     */
+    class DocIdsIteratorValues implements KNNVectorValuesIterator {
+        protected DocIdSetIterator docIdSetIterator;
+
+        DocIdsIteratorValues(@NonNull final DocIdSetIterator docIdSetIterator) {
+            this.docIdSetIterator = docIdSetIterator;
+        }
+
+        @Override
+        public int docId() {
+            return docIdSetIterator.docID();
+        }
+
+        @Override
+        public int advance(int docId) throws IOException {
+            return docIdSetIterator.advance(docId);
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return docIdSetIterator.nextDoc();
+        }
+
+        @Override
+        public DocIdSetIterator getDocIdSetIterator() {
+            return docIdSetIterator;
+        }
+
+        @Override
+        public long liveDocs() {
+            if (docIdSetIterator instanceof BinaryDocValues) {
+                return KNNCodecUtil.getTotalLiveDocsCount((BinaryDocValues) docIdSetIterator);
+            } else if (docIdSetIterator instanceof FloatVectorValues || docIdSetIterator instanceof ByteVectorValues) {
+                return docIdSetIterator.cost();
+            }
+            throw new IllegalArgumentException(
+                "DocIdSetIterator present is not of valid type. Valid types are: BinaryDocValues, FloatVectorValues and ByteVectorValues"
+            );
+        }
+    }
+
+    /**
+     * A FieldWriterIteratorValues is mainly used when Vectors are stored in {@link KnnFieldVectorsWriter} interface.
+     */
+    class FieldWriterIteratorValues<T> implements KNNVectorValuesIterator {
+        private final DocIdSetIterator docIdSetIterator;
+        private final Map<Integer, T> vectors;
+
+        FieldWriterIteratorValues(@NonNull final DocsWithFieldSet docsWithFieldSet, @NonNull final Map<Integer, T> vectors) {
+            assert docsWithFieldSet.iterator().cost() == vectors.size();
+            this.vectors = vectors;
+            this.docIdSetIterator = docsWithFieldSet.iterator();
+        }
+
+        @Override
+        public int docId() {
+            return docIdSetIterator.docID();
+        }
+
+        @Override
+        public int advance(int docId) throws IOException {
+            return docIdSetIterator.advance(docId);
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return docIdSetIterator.nextDoc();
+        }
+
+        /**
+         * Returns a Map of docId and vector.
+         * @return {@link Map}
+         */
+        public T vectorsValue() {
```

Comment:
I like this, was wondering if we can have something similar for `DocIdsIteratorValues`, we probably will need separate implementations for float and byte there but might simplify the logic in KNNVectorValues end

Reply:
@shatejas in terms of logic simplification, there are already some good suggestion added, let me try that first and how code comes out. 
on your suggestion I was in a lot of dilemma whether to give Iterator the interface of getVector and I decided not to because I wanted iterator to just iterating over the docIds. But I can give another thought on this.